File: Leetcode/25_Reverse Nodes in k-Group/c++/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include <exception>

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};
  
class Solution {
public:
    int lengthOfList(ListNode* head) {
        int length = 0;
        while(head != nullptr) {
            length++;
            head = head -> next;
        }
        
        return length;
    }
    
    
    ListNode* reverseKGroup(ListNode* head, int k) {
        if(head == nullptr)
            return nullptr;
        
        if(lengthOfList(head) < k)
            return head;
        
        ListNode* prev = nullptr;
        ListNode* current = head;
        ListNode* reversedHead = nullptr;
        for(int i = 0; i < k; i++) {
            ListNode* pNext = current -> next;
                
            current -> next = prev;
            
            prev = current;
            current = pNext;   
        }
        head -> next = reverseKGroup(current, k);
        
        return prev;
    }
};

File: Leetcode/25_Reverse Nodes in k-Group/python/solution.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def reverseLinkedList(self, node, previousNode, k):
        start = node
        t1 = node
        t2 = node.next
        nextNode = None
        
        while k > 0 and t2:
            nextNode = t2.next
            t2.next = t1
            t1 = t2
            t2 = nextNode
            k -= 1

        start.next = nextNode
        if previousNode:
            previousNode.next = t1
        if not self.finalStartNode:
            self.finalStartNode = t1
        
        return nextNode, start

    def getLengthOfLinkedList(self, head):
        count = 0
        while head:
            count += 1
            head = head.next
        return count

    def reverseKGroup(self, head, k):
        if not head or k == 1:
            return head

        previousNode = None 
        self.finalStartNode = None
        lengthOfLists = self.getLengthOfLinkedList(head)
        
        if k > lengthOfLists:
            return head

        numberOfIterations = int(lengthOfLists / k)
        while head and numberOfIterations > 0:
            newStartNode, newEndNode = self.reverseLinkedList(head, previousNode, k-1)
            previousNode = newEndNode
            head = newStartNode
            numberOfIterations -= 1

        return self.finalStartNode
        

        

